### DISCUSSION TITLE
Docker containers

### DISCUSSION THREAD

**Discussion by AnkurSheel** [BlazingSaddlesSolutions/TimeKeeper#63] [General]
I have been playing around with dockerizing our application and had a couple of questions for the community

- Do we want the different components (db, API, client, client-server) on specific ports?.
- Should we update the folder structure and move the code/tests into a sub-folder?
- Do we want separate docker-compose files for spinning up the client and client-server applications?

If anyone wants to play around with it, I have pushed up a WIP branch ([docker_containers](https://github.com/BlazingSolutions/TimeKeeper/tree/docker_containers)) which spins up the db on port 1433 and blazor client on port 8080.

**Comment by DotNetDublin** [comment]
Thanks @AnkurSheel, this is great.

> Do we want the different components (db, API, client, client-server) on specific ports?

Ports came up in this [discussion](https://github.com/BlazingSolutions/TimeKeeper/discussions/58#discussioncomment-770337) and you'll note [TimeKeeper.Api/Startup.cs](https://github.com/BlazingSolutions/TimeKeeper/blob/main/TimeKeeper.Api/Startup.cs) references ports in relations to it CORS policy.

> Should we update the folder structure and move the code/tests into a sub-folder?

Would that be where all the projects are moved into a folder named src or similar?

Not sure if anyone would have an issue with that however perhaps you could confirm the folder structure you had in mind and the benefits you see this bringing.

> Do we want separate docker-compose files for spinning up the client and client-server applications?

I'll leave that for those more experienced with Docker however do you see advantages in having a separate one?

> If anyone wants to play around with it, I have pushed up a WIP branch (docker_containers) which spins up the db on port 1433 and blazor client on port 8080.

Is there an article or YouTube tutorial you would recommend for those of us new to Docker on how to get started with what's in the docker_containers branch?

**Reply by AnkurSheel** [reply]
> Ports came up in this discussion and you'll note TimeKeeper.Api/Startup.cs references ports in relations to it CORS policy.

Cool. Will take a look and try to keep it the same. 

> Would that be where all the projects are moved into a folder named src or similar?

Yup. 
A similar structure to below has worked well for me but 
TimeKeeper
  -> build -> Dockerfiles etc
  -> src -> code
  -> buildScripts -> deployment scripts
  -> infrastructure -> terraform scripts 

> Is there an article or YouTube tutorial you would recommend for those of us new to Docker on how to get started with what's in the docker_containers branch?

Chris Sainity has some good articles on setting up [blazor with docker](https://chrissainty.com/containerising-blazor-applications-with-docker-containerising-a-blazor-webassembly-app/).
But a lot of what I have built so far has been from all over the place. I will see if I get time to write something up over the weekend. Is there anything in particular that you are interested in?

**Reply by DotNetDublin** [reply]
Thanks.

I wouldn't be upset over a structure like the one you propose however we might open a separate discussion on it in case someone has a suggestion on it.

I'll check out the Chris Sainty article. I guess what I had in mind is something similar to our "Getting Started" in that we'd say to run the application within Docker perform the following steps:

- Install Docker
- Enter the below command line

I'll probably have better points once I've gone through Chris's article.
